chore: remove `bulletproofs` and use unblinded amounts

RolandSherwin opened the pull request.

RolandSherwin committed:
chore: remove `bulletproofs` and use unblinded amounts

grumbach commented on src/transaction/mod.rs:
```diff
@@ -131,19 +102,9 @@ impl DbcTransaction {
-            .map(BlindedOutput::blinded_amount)
-            .map(RistrettoPoint::from)
-            .sum();
+        // Check that the input and output amounts are equal.
+        let input_sum: u64 = self.inputs.iter().map(|i| i.amount.value).sum();
+        let output_sum: u64 = self.outputs.iter().map(|o| o.amount.value).sum();
 
```
We might want to check for overflows here, on both inputs and outputs. 

```
input = 10
output1 = U64_MAX
output2 = 11

input == output1 + output2
```

In the above scenario, transaction is balanced!

grumbach replied:
Just checked, RUST panics in that case! We're safe! But we might want to check so our program doesn't crash!

RolandSherwin replied:
Aha! Thanks for pointing that out, will use `checked_add` here!

grumbach commented on src/transaction/mod.rs:
```diff
@@ -153,195 +114,56 @@ impl DbcTransaction {
-    input_amounts: &[BlindedAmount],
-    blinded_outputs: &[BlindedOutput],
-) -> Vec<u8> {
+fn serialize(dbc_ids: &[DbcId], input_amounts: &[Amount], outputs: &[Output]) -> Vec<u8> {
```
I think me might not need a manual `serialize` for DBCs anymore as we don't have the un-serializable types from bulletproofs, we can just derive the trait.

grumbach commented on src/transaction/input.rs:
```diff
@@ -73,19 +33,34 @@ impl BlindedInput {
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone)]
+pub struct InputIntermediate {
```
Just brainstorming here, any chance we could get rid of this extra type and only have one type for `Input` in the API? 
I would simplify our exposed API as we'd only have one input type. 
Just wondering here, and also probably out of scope for this PR.

RolandSherwin replied:
Removing `Input::signature` might actually allow us to remove this extra struct. And I think it would be safe to remove the `signature` field? It has been briefed in the above comment.

grumbach reviewed (commented):
Great one Roland! One note, I think we don't need Amounts in ciphers anymore, and we can also remove all custom serialisation (`to_bytes`) as we got rid of the bulletproof types that prevented us from deriving `serde` traits.

RolandSherwin commented on src/builder.rs:
```diff
@@ -198,7 +172,7 @@ impl DbcBuilder {
     /// See TransactionVerifier::verify() for a description of
     /// verifier requirements.
-    pub fn build(self) -> Result<Vec<(Dbc, RevealedAmount)>> {
+    pub fn build(self) -> Result<Vec<(Dbc, Amount)>> {
```
Ah, @bochaco had suggested me earlier that we might still have the amount there because it might be convenient for the users.

grumbach replied:
Even if we have `dbc.amount()` surely that's convenient enough for the users?
My concern is that one more copy of the same data means one more to check and keep track of to make sure the value is identical. In general I'd be against returning some data twice. @bochaco what's your take on this?

bochaco replied:
ahh, yeah when I suggested that we were still encrypting the amounts, so the user would have had to decrypt them, now they are not, so either way works fine for the user

RolandSherwin committed:
chore(deps): removed unused dependencies

RolandSherwin commented on src/builder.rs:
```diff
@@ -208,60 +182,35 @@ impl DbcBuilder {
+                    .get(&output.dbc_id)
+                    .ok_or(Error::DbcIdNotFound)?;
                 let ciphers = DbcCiphers::from((
                     &dbc_id_src.public_address,
                     &dbc_id_src.derivation_index,
-                    revealed_amounts[0],
+                    output.amount,
                 ));
```
Ah! I see, overlooked that one. The `SignedSpend` would contain the signature over the `Amount` as well, so this would not be necessary anymore.

RolandSherwin committed:
chore: remove `Amount` cipher from `DbcCiphers`

RolandSherwin commented on src/transaction/input.rs:
```diff
@@ -4,67 +4,27 @@
     pub dbc_id: DbcId,
-    pub blinded_amount: BlindedAmount,
+    pub amount: Amount,
     /// This is the signature of the `DerivedKey`
     /// corresponding to this `dbc_id`
     pub signature: Signature,
```
It is the signature over the `input dbc_ids`, `input amount` and the `Vec<Output>`. 
i.e, it signs the message that we get from `fn serialize(dbc_ids: &[DbcId], input_amounts: &[Amount], outputs: &[Output]) -> Vec<u8>`

I assume we might not need it anymore? Since the `SignedSpend` has the signature over everything? 
If so, this might allow us to get rid of the `InputInteremediate` struct.

grumbach replied:
Yes this is weird, we create a signature ourselves and then check it ourselves without the Network or anybody else ever seeing it... An artefact from the time when this job was done by elders and not clients?
If we have a sig in SignedSpend then this can go indeed.

RolandSherwin replied:
Neat, thanks for confirming it! This will simplify a lot of things!

grumbach replied:
> It is the signature over the input dbc_ids, input amount and the Vec<Output>.
Could you add that description in the comment?

RolandSherwin committed:
fix: check for overflow while adding values

RolandSherwin committed:
chore: remove redundant `Input::signature`
- The dbc is signed to obtain the `SignedSpend`, so this signature can be removed

RolandSherwin committed:
chore(ci): remove auto merge workflow
